src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,144 @@
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Objects.checkIndex(index, size + 1);
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index - 1);
+            newNode.next = nodeByIndex.next;
+            nodeByIndex.next.prev = newNode;
+            nodeByIndex.next = newNode;
+            newNode.prev = nodeByIndex;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            throw new NullPointerException("This list is empty" + list);
+        }
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> newNode = getNodeByIndex(index);
+        T oldElement = newNode.element;
+        newNode.element = value;
+        return oldElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        T removedElement = getRemovedElement(index);
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            throw new NoSuchElementException();
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < size; i++) {
+                if ((current.element != null && current.element.equals(object))
+                        || (current.element == null && object == null)) {
+                    remove(i);
+                    return true;
+                }
+                current = current.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
```

Comment:
This method may take a lot of memory. So it is asked in checklist to split it into 2 loops - first half of the list and second. Logic is right.